## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [3]:
# Use the describe function to gather some basic statistics

description = df.describe()
print(description)

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#data types
print(df.dtypes)

#Rows and Columns
print(df.shape)
#Types and Memory
print(df.info())
#Missing Values
print(df.isnull().sum())
#Unique Values
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Unique values in {column}: {unique_values}")


first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object
(54639, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-nul

In [5]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)

print("Three item categories with the most entries:")
print(top_categories)

Three item categories with the most entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [6]:
# For the category with the most entries,
# which subcategory had the most entries?

# Most entries
most_entries_category = df['category'].value_counts().idxmax()

filtered_df = df[df['category'] == most_entries_category]

# Count
most_entries_subcategory = filtered_df['subcategory'].value_counts().idxmax()

# Subcategory
print(f"The subcategory with the most entries within the category '{most_entries_category}' is '{most_entries_subcategory}'.")



The subcategory with the most entries within the category 'consumables' is 'bathroom supplies'.


In [7]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
print("Five clients with the most entries:")
print(top_clients)


Five clients with the most entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [8]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5)
top_clients_list = top_clients.index.tolist()

print("Client IDs of the top 5 clients:")
print(top_clients_list)


Client IDs of the top 5 clients:
[33615, 66037, 46820, 38378, 24741]


In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?

# most entries
most_entries_client = df['client_id'].value_counts().idxmax()

filtered_df = df[df['client_id'] == most_entries_client]

# Sum
total_units_ordered = filtered_df['qty'].sum()

print(f"The client with the most entries ordered a total of {total_units_ordered} units.")

The client with the most entries ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

# New column 'subtotal'
df['subtotal'] = df['unit_price'] * df['qty']

# Print
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def calculate_shipping_price(unit_weight):
    if unit_weight > 50:
        return 7 * unit_weight
    else:
        return 10 * unit_weight
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(calculate_shipping_price)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)




unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

def calculate_total_price(subtotal, shipping_price, tax_rate):
    sales_tax = (subtotal + shipping_price) * (tax_rate / 100)
    total_price = subtotal + shipping_price + sales_tax
    return total_price

# Sales tax rate of 9.25%
sales_tax_rate = 9.25 / 100

# New column 'total_price'
df['total_price'] = df.apply(lambda row: calculate_total_price(row['subtotal'], row['shipping_price'], sales_tax_rate), axis=1)

df["total_price"] = df["total_price"].round(2)


# print
df[['shipping_price', 'subtotal', 'total_price']].head(3)



shipping_price   subtotal  total_price
0         5512.50  115164.00    120788.13
1          312.90     523.95       837.62
2          458.64     527.28       986.83

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

# New column 'line_cost'
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

# Print
print(df.head())

       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.5

In [14]:
# Create a column for the profit of
# each line using line cost and line price

# Line price
df['line_price'] = df['unit_price'] * df['qty']

# Line cost
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

# New column 'line_profit'
df['line_profit'] = df['line_price'] - df['line_cost']

# Print
print(df.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1 

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
# Calculate line price
df['line_price'] = df['unit_price'] * df['qty']

# Calculate line cost
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']

# Calculate line profit
df['line_profit'] = df['line_price'] - df['line_cost']

# Include only orders 2742071, 2173913, and 6128929
filtered_df = df[df['order_id'].isin([2742071, 2173913, 6128929])]

# Group by order numbers
order_totals = filtered_df.groupby('order_id').agg(
    total_line_price=('line_price', 'sum'),
    total_line_cost=('line_cost', 'sum'),
    total_line_profit=('line_profit', 'sum')
)

# Print
print(order_totals)

          total_line_price  total_line_cost  total_line_profit
order_id                                                      
2173913          109440.76        130500.17          -21059.41
2742071          127799.33        124058.01            3741.32
6128929          688564.27        625740.71           62823.56


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# Group  by client ID
client_spending = df.groupby('client_id').agg(
    total_spending=('subtotal', 'sum'),
    total_shipping=('shipping_price', 'sum')
)

# Total spending 
client_spending['total_spending_including_shipping'] = client_spending['total_spending'] + client_spending['total_shipping']

# Top 5 clients
top_clients_by_quantity = df['client_id'].value_counts().head(5)

top_clients_spending = client_spending.loc[top_clients_by_quantity.index]

print(top_clients_spending)

           total_spending  total_shipping  total_spending_including_shipping
client_id                                                                   
33615          5839032.11      1828984.89                         7668017.00
66037          7995708.38      1395151.85                         9390860.23
46820          7317356.03      1601448.84                         8918804.87
38378          8384321.15      3429455.40                        11813776.55
24741         70176885.25      5126448.37                        75303333.62


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# Group by client ID
client_summary = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('subtotal', 'sum'),
    total_profit=('line_profit', 'sum')
)

# Top 5 clients
top_clients = client_summary.nlargest(5, 'total_profit')

# Print Summary
print(top_clients)

           total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
78965                    4033302            5340756.88   5.950470e+08   
85931                     644423           10122549.19   2.071928e+08   
24741                     239862            5126448.37   7.017689e+07   
38473                     402770            4391003.73   1.636205e+08   
84344                     466968            4746583.13   9.287183e+07   

           total_profit  
client_id                
78965      1.864762e+08  
85931      7.906991e+07  
24741      2.448799e+07  
38473      2.341966e+07  
84344      1.393541e+07  


In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

# Calculate total revenue and total profit
df['total_revenue'] = df['subtotal'] + df['shipping_price']
df['total_profit'] = df['total_revenue'] - df['line_cost']

# Convert currency columns to millions of dollars
df['total_revenue_million'] = df['total_revenue'] / 1e6
df['total_profit_million'] = df['total_profit'] / 1e6

# Rename columns
df.rename(columns={
    'client_id': 'Client ID',
    'qty': 'Total Units Purchased',
    'shipping_price': 'Total Shipping Price (Millions of Dollars)',
    'total_revenue_million': 'Total Revenue (Millions of Dollars)',
    'total_profit_million': 'Total Profit (Millions of Dollars)'
}, inplace=True)

formatted_df = df[['Client ID', 'Total Units Purchased', 'Total Shipping Price (Millions of Dollars)', 
                   'Total Revenue (Millions of Dollars)', 'Total Profit (Millions of Dollars)']]

# Print
print(formatted_df)





       Client ID  Total Units Purchased  \
0          58515                    105   
1          37609                     21   
2          57113                     39   
3          46554                     29   
4          92089                     20   
...          ...                    ...   
54634      29043                     33   
54635      41908                     47   
54636      35176                    475   
54637      24485                    112   
54638      24007                   1031   

       Total Shipping Price (Millions of Dollars)  \
0                                         5512.50   
1                                          312.90   
2                                          458.64   
3                                          356.70   
4                                         6500.20   
...                                           ...   
54634                                      519.75   
54635                                     3849.30   
54636 

In [36]:
# Sort the updated data by "Total Profit" form highest to lowest
print(df.columns)


# Sort the DataFrame by "Total Profit (Millions of Dollars)" from highest to lowest
df_sorted = df.sort_values(by='Total Profit (Millions of Dollars)', ascending=False)

# Calculate Total Cost as the difference between Total Revenue and Total Profit
df_sorted['Total Cost (Millions of Dollars)'] = df_sorted['Total Revenue (Millions of Dollars)'] - df_sorted['Total Profit (Millions of Dollars)']

# Format the columns
df_sorted['Total Revenue (Millions of Dollars)'] = df_sorted['Total Revenue (Millions of Dollars)'].apply(lambda x: f'${x:.2f}M')
df_sorted['Total Cost (Millions of Dollars)'] = df_sorted['Total Cost (Millions of Dollars)'].apply(lambda x: f'${x:.2f}M')
df_sorted['Total Profit (Millions of Dollars)'] = df_sorted['Total Profit (Millions of Dollars)'].apply(lambda x: f'${x:.2f}M')

# Keep only the desired columns
df_sorted = df_sorted[['Client ID', 'Total Units Purchased', 'Total Shipping Price (Millions of Dollars)',
'Total Revenue (Millions of Dollars)', 'Total Cost (Millions of Dollars)', 'Total Profit (Millions of Dollars)']]


# Display only the top 3 entries
print(df_sorted.head(5))

print('''I had to get lots of help on this one and still think I got it mostly wrong.
I had problems with my totals in multiple sections.  Will reach out for help.''')

Index(['first', 'last', 'job', 'phone', 'email', 'Client ID', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight',
       'Total Units Purchased', 'line_number', 'subtotal', 'total_weight',
       'Total Shipping Price (Millions of Dollars)', 'total_price',
       'line_cost', 'line_price', 'line_profit', 'total_revenue',
       'total_profit', 'Total Revenue (Millions of Dollars)',
       'Total Profit (Millions of Dollars)'],
      dtype='object')
       Client ID  Total Units Purchased  \
21801      78965                3958244   
15112      85931                 637318   
25972      86721                 977121   
4063       24741                 122166   
23929      38473                 219645   

       Total Shipping Price (Millions of Dollars)  \
21801                                  2493693.72   
15112                                  9859309.46   
25972                            